# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/vr/ipc/VRManagerChild.cpp
# Commit: cc83c2684cbb
# Full Hash: cc83c2684cbb2ee2b8b6c1f8f3e0dd3a5b586ac0
# Author: sotaro <sotaro.ikeda.g@gmail.com>
# Date: 2020-01-31 21:33:33
# Regressor Bug: 1612256
# File Overlap Count: 2
# Description:
#   Bug 1612256 - Change VRManagerChild::ShutDown() similar to CompositorBridgeChild::ShutDown() r=nical
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D61263
# ==============================================================================

diff -r 846c54e66e4e -r cc83c2684cbb gfx/vr/ipc/VRManagerChild.cpp
--- a/gfx/vr/ipc/VRManagerChild.cpp	Fri Jan 31 08:27:14 2020 +0000
+++ b/gfx/vr/ipc/VRManagerChild.cpp	Fri Jan 31 08:20:56 2020 +0000
@@ -137,23 +137,22 @@
   MOZ_ASSERT(NS_IsMainThread());
   if (sVRManagerChildSingleton) {
     sVRManagerChildSingleton->Destroy();
-    sVRManagerChildSingleton = nullptr;
+    SpinEventLoopUntil([&]() { return !sVRManagerChildSingleton; });
   }
 }
 
-/*static*/
-void VRManagerChild::DeferredDestroy(RefPtr<VRManagerChild> aVRManagerChild) {
-  aVRManagerChild->Close();
-}
-
 void VRManagerChild::Destroy() {
   // Keep ourselves alive until everything has been shut down
   RefPtr<VRManagerChild> selfRef = this;
+  MessageLoop::current()->PostTask(NewRunnableMethod(
+      "VRManagerChild::AfterDestroy", selfRef, &VRManagerChild::AfterDestroy));
+}
 
-  // The DeferredDestroyVRManager task takes ownership of
-  // the VRManagerChild and will release it when it runs.
-  MessageLoop::current()->PostTask(NewRunnableFunction(
-      "VRManagerChildDestroyRunnable", DeferredDestroy, selfRef));
+void VRManagerChild::AfterDestroy() {
+  Close();
+  if (sVRManagerChildSingleton == this) {
+    sVRManagerChildSingleton = nullptr;
+  }
 }
 
 PVRLayerChild* VRManagerChild::AllocPVRLayerChild(const uint32_t& aDisplayID,